zerver/migrations/0250_fix_has_link_attribute.py

```diff
@@ -0,0 +1,48 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.24 on 2019-10-07 05:25
+from __future__ import unicode_literals
+
+from django.db import migrations
+from django.db.backends.postgresql_psycopg2.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+import lxml
+import urllib
+
+def fix_has_link(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    Message = apps.get_model('zerver', 'Message')
+    Attachment = apps.get_model('zerver', 'Attachment')
+    for message in Message.objects.all():
+        lxml_obj = lxml.html.fromstring(message.rendered_content)
+        has_link = False
+        has_attachment = False
+        has_image = False
+
+        for link in lxml_obj.xpath("//a"):
+            has_link = True
+            href = link.attrib["href"]
+            parsed_href = urllib.parse.urlsplit(href)
+            if not parsed_href.path.startswith("/user_uploads/"):
+                continue
+            path_id = parsed_href.path[len("/user_uploads/"):]
+            if Attachment.objects.filter(path_id=path_id).exists():
+                has_attachment = True
```

Comment:
This might be done better by just querying the `AttachmentMessage` table for the message; that should agree with this parsing logic.

Reply:
To investigate whether we can just do `has_attachment = message.attachment_set.exists()`, I'm running the following query on chat.zulip.org and got not output other than the progress reports to look for anomalies where those disagree:

```
for message in Message.objects.all(): 
    if message.id % 1000 == 0: 
        print(message.id) 
    lxml_obj = lxml.html.fromstring(message.rendered_content) 
    has_attachment = False 
    for link in lxml_obj.xpath("//a"): 
        href = link.attrib["href"] 
        parsed_href = urllib.parse.urlsplit(href)  
        if not parsed_href.path.startswith("/user_uploads/"): 
            continue 
        path_id = parsed_href.path[len("/user_uploads/"):] 
        attachment = Attachment.objects.filter(path_id=path_id).first() 
        if attachment is not None and message.id not in [m.id for m in attachment.messages.all()]: 
            print(message.id, [m.id for m in attachment.messages.all()]) 
        elif attachment is not None: 
            has_attachment = True 
    if not has_attachment: 
        if len(message.attachment_set.all()) == 0: 
            continue 
        print([a.path_id for a in message.attachment_set.all()]) 
```